lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,296 @@
+                reverse=True
+            )
+            bands = bands[:nr_bands]
+            bands = sorted(bands, key=itemgetter(0))
+
+        return bands
+
+    @classmethod
+    def _get_phases(cls, plat_info):
+        capacities = cls._collect_capacities_flatten(plat_info)
+        bands = cls._get_bands(capacities)
+
+        def _band_mid(band):
+            return int((band[1] + band[0]) / 2)
+
+        """
+        Returns a list of phases. Each phases is described by a tuple:
+
+          (uclamp_val, util, (band))
+        """
+        return [(_band_mid(band),
+                 _band_mid(band) / 2,
+                 band)
+                for index, band in enumerate(bands)]
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        periods = [
+                Periodic(
+                    duty_cycle_pct=(util / PELT_SCALE) * 100,  # util to pct
+                    duration_s=5,
+                    period_ms=cls.TASK_PERIOD_MS,
+                    uclamp_min=uclamp_val,
+                    uclamp_max=uclamp_val)
+                for uclamp_val, util, expected in cls._get_phases(plat_info)
+        ]
+
+        return {'task': functools.reduce(lambda a, b: a + b, periods)}
+
+        return df
+
+    def _get_task_activation_capacity(self, task_name, start, end):
+
+        df_task = self.trace.analysis.tasks.df_task_activation(task_name)
+        df_task = df_task[['active', 'cpu']]
+
+        df_freq = self.trace.analysis.frequency.df_cpus_frequency()
+        df_freq = df_freq[['cpu', 'frequency']]
+        df_freq = df_freq.pivot(index=None,
+                                columns='cpu',
+                                values='frequency').reset_index()
+        df_freq.set_index('Time', inplace=True)
+
+        df = pd.concat([df_task, df_freq])
+        df = df.sort_index()
+
+        # Find the first activation
+        start = df[(df.index >= start) & (df.active == 1)].first_valid_index()
+
+        if (start >= end):
+            raise ValueError('Phase ends before it has even started')
+
+        df = df.ffill()
+        df = df[df.index >= start]
+        if (not np.isnan(end)):
+            df = df[df.index <= end]
+
+        # Fake Time, representing only the task activation period
+        df_add_delta(df, col='Time', inplace=True)
+        df = df.dropna()
+        df = df[df.active == 1]
+        df['Time'] = df['Time'].cumsum()
+
+        # Frequency to capacity
+        capacities = self._collect_capacities(self.plat_info)
+        for cpu in self.plat_info['nrg-model'].cpu_nodes:
+            cpu_id = cpu.cpus[0]
+            df[cpu_id] = df[cpu_id].map(capacities[cpu_id])
+
+        df['capacity'] = df.apply(lambda x: x[x.cpu], axis=1)
+        df['capacity_max'] = df.apply(
+                lambda x: max(capacities[x.cpu].values()), axis=1)
+
+        return [round(series_mean(df['capacity'], x=df['Time'])),
+                round(series_mean(df['capacity_max'], x=df['Time']))]
```

Comment:
This is to give the test a chance to succeed if the task needs to down migrate.

Reply:
If we expect migrations to happen within a phase, we should check for it and ignore what happens before, averaging is not something we want to end up doing for that purpose. In the past 2 years we (painfully) scrubbed LISA to eliminate as much of that sort of margin as possible because they are:
* never computed from a knowledge model, it's always empirical
* from the above follows:
    * no-one can understand them, because usually no-one ever did, including the original test writer. You might think you know why you added it, but the reality is that you could be subject to an rt-app bug or some CPU-invariance issue and you would not realize, unless you are digging to check a specific hypothesis.
    * they simply fail randomly on some devices, because they amounted to essentially hardcoding some characteristics that should be derived from PlatformInfo
* hiding real issues (both in the kernel and in the test)

The last point can escalate to a much worse issue: a test appears to be simple, and check an "obvious" property of a kernel's signal. Turns out the signal does not actually have the property to start with, so the test should actually be expected to fail, but it passed because margins were added to "accommodate" the situation (test_util_migration is a prime example of how a "high level" property about the sum of 2 util signals actually does not hold, and the real property is significantly more tricky to check).

So essentially, all new tests must be designed as:
* no margins (or only on last resort, for like 1 or 2% on an absolute scale)
* reimplement the kernel bit you want to test in Python
* expose the inputs and outputs of the kernel bit, and check against the Python implementation

When the test fails, we know exactly where, and it's much more straightforward to understand why. Also it works on all platforms, and when it does not, it's usually easy to understand why and the fix is basically just exposing some kernel state that was hidden.